Use fs.realpathSync.native when available

amcasey opened the pull request:
In local measurements of an Office project, we saw initial project loading get 5% faster on Windows and 13% faster on Linux.  The only identified behavioral change is that it restores the case used on disk, whereas realpathSync retains the input lowercase.

amcasey commented:
I did local validation of this as I was testing https://github.com/microsoft/TypeScript/pull/41288.  It was less effective, so I dropped it, but it has the advantage of handling directory symlinks correctly, which that PR does not (at time of writing).

amcasey commented:
@typescript-bot perf test this

amcasey commented:
We can't characterize the differences between `realpathSync` and `realpathSync.native` or the behavior we expect from either, so we're not ready to make this change.

amcasey commented:
The risk seems more acceptable during the 4.2 beta period.  The differences I have uncovered so far are that, on Windows, the native implementation will return the letter-casing found on disk and will leave network paths intact, even if they resolve to the current machine.

amcasey commented:
It looks like we have a common pattern of calling `realpath`, re-normalizing the slashes, and then comparing that value to the original path.  This depends on `realpath` (arguably, incorrectly) not matching the letter-casing found on disk.

amcasey commented:
I am astonished to learn that this behavior is [documented](https://nodejs.org/api/fs.html#fs_fs_realpath_path_options_callback):

> No case conversion is performed on case-insensitive file systems.

sheetalkamat commented:
As discussed offline we should have some environment specific tests for hosts where `realpath` returned differs in casing etc. 
i did verify all locations of `realpath` calls and it seems we check casing correctly so there shouldn't be any issues in comparing realpath.. but as usual worry is `forceConsistentCasingInFileNames` which needs some tests to ensure its working ok.

sandersn commented:
@amcasey this doesn't have an associated bug -- are you aiming to ship this in 4.2?

amcasey commented:
@sandersn Yes, I'm hoping to get it in before the beta because we're pretty uncertain of the consequences.  I'm currently trying some testing strategies Sheetal suggested.

amcasey commented:
I added some new baseline tests of `forceConsistentCasingInFileNames` based on an offline discussion with @sheetalkamat and then updated both the VFS and the VFSWithWatch to use the revised realpath semantics.  Only one existing baseline changed and it was a test issue, so I updated the test (to a state where it passes with both versions of realpath).  I did not retain testing of the old semantics - even though they'll still be used as a fallback - because duplicating all of the tests seemed excessive.

amcasey commented:
@typescript-bot test this

amcasey commented:
@typescript-bot perf test this

amcasey commented:
@typescript-bot user test this

amcasey commented:
@typescript-bot perf test this

amcasey commented:
Compiler-Unions emit went from +3% in the first run to -3% in the second run, so I'm going to assume it's just noisy.

amcasey commented:
I ran the RWC tests locally on a dummy change, updated all the baselines, and then ran them again on this change - there were no further baseline changes.  I will attempt to do something similar for the User Tests.

rbuckton commented:
In the NodeJS docs there is this note regarding `realpathSync.native`:

> ![image](https://user-images.githubusercontent.com/3902892/102546115-00437a80-406c-11eb-9b13-66444df8aaa3.png)

Do we know if that means the function will throw for the case above, or won't exist?

amcasey commented:
@rbuckton Do you have a link?  I remember reading about some compiler issues, but I thought it only affected people building their own because the official drops were built with a setup that would make things work.  Having said that, no, I have no idea what the failure mode is in that case.  It seems like not existing is how you would want it to manifest, but that may not be what actually happens.

I did, FWIW, test a prototype of this change on ubuntu server.

rbuckton commented:
Also, running perf tests on Windows can be noisy. There are two things to do for more stable perf numbers:

- If you're using a laptop make sure your power mode is set to High Performance, otherwise Windows will throttle the CPU in certain cases
- Disable Windows Defender _during a perf run_, or _temporarily_ add NodeJS and your TypeScript directory to the Windows Defender exclusion list so that I/O isn't hampered by AV:
  - Open the Windows Security app (or go to Setings, Update & Security, Windows Security and click the "Open Windows Security" button to open the app)
  - In the Windows Security app, click the "Virus & threat protection" menu on the left.
  - Under "Virus & threat protetion settings" click "Manage settings"
  - Scroll down to the "Exclusions" section and click "Add or remove exclusions"
  - Add the NodeJS executable and your TypeScript directory as excluded paths.
  - NOTE: To be safe, after a perf run you should remove the exclusions.

Its possible to script the above steps using PowerShell (or at least, it used to be). If you're interested ping me on Teams.

rbuckton commented:
> @rbuckton Do you have a link? ...

It's in the documentation for `realPathSync.native`, in the last paragraph: https://nodejs.org/dist/latest-v15.x/docs/api/fs.html#fs_fs_realpathsync_native_path_options

amcasey commented:
@rbuckton I think musl libc is an incredibly small niche, but I can try to set up a VM if you're concerned.

I didn't use the official perf run to measure the performance - that was just a sanity check.  The actual measurements were from a harness driving tsserver and tracked the updateOpen time across 9 or so runs.  I tested on both Windows and Linux.  I can try to dig up the original numbers, if you're interested.  Anecdotally, when I was profiling midgard yesterday, I saw that realpathSync took 20% of total updateOpen time, whereas realpathSync.native took only 10% (on windows).

amcasey commented:
Windows|||||||||||||
-|-|-|-|-|-|-|-|-|-|-|-|-|
baseline|Initialization|289|290|255|256|300|263|261|261|274|270|100%|
baseline|OpenProject|7577|7967|7470|7451|7528|7412|7480|7487|7418|7527|100%|
baseline|Synchronize|11|9|8|10|9|8|9|8|8|9|100%|
native|Initialization|284|260|296|255|261|287|304|277|262|275|102%|
native|OpenProject|7181|7163|7058|7080|7151|7297|7147|7199|7257|7169|95%|
native|Synchronize|9|9|9|10|9|9|9|9|9|9|106%|


Linux|||||||||||||
-|-|-|-|-|-|-|-|-|-|-|-|-
baseline|Initialization|182|183|189|183|189|184|185|184|185|185|100%
baseline|OpenProject|9505|9533|9407|9514|9573|9536|9574|9749|9533|9552|100%
baseline|Synchronize|34|34|35|34|35|34|35|34|35|35|100%
native|Initialization|175|210|183|184|184|184|183|184|185|187|101%
native|OpenProject|8452|8336|8105|8154|8321|8487|8294|8397|8377|8309|87%
native|Synchronize|34|34|34|34|34|35|34|35|34|34|99%

amcasey commented:
musl-based Alpine Linux running in RAM (notes suggested ramdisk was the most problematic case):
![image](https://user-images.githubusercontent.com/10508071/102555870-849df980-407c-11eb-844d-7e6a8cf0335b.png)

amcasey commented:
@typescript-bot user test this

amcasey commented:
If no one screams, I'm merging this tomorrow.  I hereby acknowledge that it will (and should) be rolled back at the first sign of trouble.

rbuckton commented on src/harness/collectionsImpl.ts:
```diff
@@ -50,6 +50,11 @@ namespace collections {
             return index >= 0 ? this._values[index] : undefined;
         }
 
+        public getKeyAndValue(key: K): [ K, V ] | [ undefined, undefined ] {
```
I'd probably call this `getEntry` to align with the `entries()` method below, though I'd make the return value `[K, V] | undefined` since you end up allocating a useless two-element array for negative cases.

amcasey replied:
Oops, missed the precedent.  Changed.

rbuckton reviewed (approved):
Minor nit, but LGTM.

amcasey commented:
After clearing up some confusion on my part, I also managed to confirm (locally, on Windows) that this change does not affect the user test baselines.

sheetalkamat commented on src/testRunner/unittests/tscWatch/forceConsistentCasingInFileNames.ts:
```diff
@@ -113,11 +113,154 @@ export const Fragment: unique symbol;
+        verifyFileSymlink("when file symlink target matches disk but import does not", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./XY`);
+        verifyFileSymlink("when import matches disk but file symlink target does not", `${projectRoot}/XY.ts`, `${projectRoot}/XY.ts`, `./Xy`);
+        verifyFileSymlink("when import and file symlink target agree but do not match disk", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./Xy`);
+        verifyFileSymlink("when import, file symlink target, and disk are all different", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./yX`);
```
I think the testcase missing that we discussed and is important change is when current directory is "c:/temp" but on disk it is "C:/Temp" and all realpath results with new API will differ in casing and give errors vs it wasn't giving earlier ?

sheetalkamat replied:
And may be it is ok for watch scenario but it is something that will impact tsserver and in turn vscode.. vscode keeps track of open files from previous session so not sure if that gets updated if you change the casing and restart vscode from directory but users will not be able to get rid of error easily unless they close all files if the casing is retained across sessions by vscode. You would need to experiment to figure that out.

amcasey replied:
> when current directory is "c:/temp" but on disk it is "C:/Temp"

I attempted to provide test coverage for drive root casing differences [here](https://github.com/microsoft/TypeScript/pull/41292/files#diff-742ed890d46f805d54cbcf33f4a6f2cd286a81ac5df890b0b20636445b464093R162).  It's entirely possible I misunderstood your concerns though - please let me know if there's something else you'd like me to test.

> may be it is ok for watch scenario but it is something that will impact tsserver

From our offline discussion (before the break), I thought I understood that the watch VFS was the best place to add these tests.  It sounds like you may be suggesting adding server tests as well.  If that's the case, can you please point me at an example I can model them on?

> vscode keeps track of open files from previous session

I think you're saying that the paths returned by the server to VS Code might reflect the casing returned by realpathSync, but I'm not sure I understand what consequences you expect from this.  If it's about re-opening files in a new session, I would have guessed VS Code used case-insensitive system calls to open them.  If it's about VS Code passing stale paths back to the server when it starts a fresh instance, I would have guessed they would be re-normalized with an additional realpathSync call.  Can you please elaborate on your concerns?

sheetalkamat replied:
> I attempted to provide test coverage for drive root casing differences here. It's entirely possible I misunderstood your concerns though - please let me know if there's something else you'd like me to test.

That test case also gives incorrect result? [This](https://github.com/microsoft/TypeScript/pull/41292/files#diff-a87a00c391bf2ad3ae44e4560646d9b8725590957f232578e15e0d6230e4477eR36) should have been earlier been error and should be error because on case sensitive file system the import will not resolve, which is what intention of `forceConsistentFileNames` is.. 
The case that it doesnt verify is also when the files on disk and all imports are consistent, but you start with current directory = different casing and build the project. Previously that would not error now it will ? 


> From our offline discussion (before the break), I thought I understood that the watch VFS was the best place to add these tests. It sounds like you may be suggesting adding server tests as well. If that's the case, can you please point me at an example I can model them on?

If the above scenario breaks for watch there is a way to fix those errors by closing tsc and restarting from correct casing directory.  But what happens in vscode is question. vscode keeps list of open files, so is the solution to close all files and open editor again? Does restart work , depending on answers to that we may need coverage for some tests here to lock the expected behaviour for baselines. There are tests for these in `unittests\tsserver\forceConsistentCasingInFileNames.ts`

amcasey replied:
I had the impression that `forceConsistentFileNames` [ignored](https://github.com/microsoft/TypeScript/blob/c456bbd4665e361a7495c3a13fd1ca9bc1265230/src/compiler/program.ts#L2533) drive roots.  Sorry for not including a comment to that effect.

> The case that it doesnt verify is also when the files on disk and all imports are consistent, but you start with current directory = different casing and build the project. Previously that would not error now it will ?

I'm not sure I follow.  Where is the current directory specified in this case?

> But what happens in vscode is question

So you'd just like me to manually verify that VS Code doesn't have obvious breaks if I close a folder, change the casing on disk, and re-open it?

sheetalkamat replied:
> I had the impression that forceConsistentFileNames ignored drive roots. Sorry for not including a comment to that effect.

Yes. But your import doesnt test import "somethingsomething/a" and  "somethingsomething/A" you would get error on either one of those and that was expected behavior which is not the current case.

> I'm not sure I follow. Where is the current directory specified in this case?

Current directory need to be specified as part of the https://github.com/microsoft/TypeScript/pull/41292/files#diff-742ed890d46f805d54cbcf33f4a6f2cd286a81ac5df890b0b20636445b464093R149 . We need test where current directory differs in casing from whats on the disk.

The bigger missing part is import in wrong casing should give error which it wont because the realpath will resolve to casing on the disk and then we will loose that info. But same code in case sensitive file system will give error because import will not be resolved.

Disk file name: a.ts 

import "a" > success
import "A" > will fail on case sensistive file systems... so it should error with forceConsistentFileCasing on case insensitive filesystem which it would not when realpath is invoked as part of this change.
